backend/drm/drm.c

```diff
@@ -0,0 +1,597 @@
+
+	if (id && *id)
+		drmModeRmFB(gbm_bo_get_fd(bo), *id);
+
+	free(id);
+}
+
+static uint32_t get_fb_for_bo(int fd, struct gbm_bo *bo)
+{
+	uint32_t *id = gbm_bo_get_user_data(bo);
+
+	if (id)
+		return *id;
+
+	id = calloc(1, sizeof *id);
+	if (!id) {
+		wlr_log(L_ERROR, "Allocation failed: %s", strerror(errno));
+		return 0;
+	}
+
+	drmModeAddFB(fd, gbm_bo_get_width(bo), gbm_bo_get_height(bo), 24, 32,
+		     gbm_bo_get_stride(bo), gbm_bo_get_handle(bo).u32, id);
+
+	gbm_bo_set_user_data(bo, id, free_fb);
+
+	return *id;
+}
+
+static bool display_init_renderer(struct wlr_drm_renderer *renderer,
+		struct wlr_drm_display *disp)
+{
+	disp->renderer = renderer;
+
+	disp->gbm = gbm_surface_create(renderer->gbm,
+				       disp->width, disp->height,
+				       GBM_FORMAT_XRGB8888,
+				       GBM_BO_USE_SCANOUT | GBM_BO_USE_RENDERING);
+	if (!disp->gbm) {
+		wlr_log(L_ERROR, "Failed to create GBM surface for %s: %s", disp->name,
+			strerror(errno));
+		return false;
+	}
+
+	disp->egl = create_platform_window_surface(renderer->egl.disp, renderer->egl.conf,
+						   disp->gbm, NULL);
+	if (disp->egl == EGL_NO_SURFACE) {
+		wlr_log(L_ERROR, "Failed to create EGL surface for %s: %s", disp->name,
+			egl_error());
+		return false;
+	}
+
+	// Render black frame
+
+	eglMakeCurrent(renderer->egl.disp, disp->egl, disp->egl, renderer->egl.context);
+
+	glViewport(0, 0, disp->width, disp->height);
+	glClearColor(0.0, 0.0, 0.0, 1.0);
+	glClear(GL_COLOR_BUFFER_BIT);
+
+	eglSwapBuffers(renderer->egl.disp, disp->egl);
+
+	struct gbm_bo *bo = gbm_surface_lock_front_buffer(disp->gbm);
+	uint32_t fb_id = get_fb_for_bo(renderer->fd, bo);
+
+	drmModeSetCrtc(renderer->fd, disp->crtc, fb_id, 0, 0,
+		       &disp->connector, 1, disp->active_mode);
+	drmModePageFlip(renderer->fd, disp->crtc, fb_id, DRM_MODE_PAGE_FLIP_EVENT, disp);
+
+	gbm_surface_release_buffer(disp->gbm, bo);
+
+	return true;
+}
+
+static drmModeModeInfo *select_mode(size_t num_modes,
+		drmModeModeInfo modes[static num_modes],
+		drmModeCrtc *old_crtc,
+		const char *str)
+{
+	if (strcmp(str, "preferred") == 0)
+		return &modes[0];
+
+	if (strcmp(str, "current") == 0) {
+		if (!old_crtc) {
+			wlr_log(L_ERROR, "Display does not have currently configured mode");
+			return NULL;
+		}
+
+		for (size_t i = 0; i < num_modes; ++i) {
+			if (memcmp(&modes[i], &old_crtc->mode, sizeof modes[0]) == 0)
+				return &modes[i];
+		}
+
+		// We should never get here
+		assert(0);
+	}
+
+	unsigned width = 0;
+	unsigned height = 0;
+	unsigned rate = 0;
+	int ret;
+
+	if ((ret = sscanf(str, "%ux%u@%u", &width, &height, &rate)) != 2 && ret != 3) {
+		wlr_log(L_ERROR, "Invalid modesetting string '%s'", str);
+		return NULL;
+	}
+
+	for (size_t i = 0; i < num_modes; ++i) {
+		if (modes[i].hdisplay == width &&
+		    modes[i].vdisplay == height &&
+		    (!rate || modes[i].vrefresh == rate))
+			return &modes[i];
+	}
+
+	wlr_log(L_ERROR, "Unable to find mode %ux%u@%u", width, height, rate);
+	return NULL;
+}
```

Comment:
Export this function and have the top level call it. If you need to pick one during init, use the preferred mode. Otherwise, make a backend-ambiguous way for the caller to set the output mode.